jvb/src/main/kotlin/org/jitsi/videobridge/dcsctp/DcSctpTransport.kt

```diff
@@ -0,0 +1,146 @@
+                put("uses_zero_checksum", metrics.usesZeroChecksum())
+                put("negotiated_maximum_incoming_streams", metrics.negotiatedMaximumIncomingStreams)
+                put("negotiated_maximum_outgoing_streams", metrics.negotiatedMaximumOutgoingStreams)
+            }
+        }
+    }
+
+    companion object {
+        private val factory = DcSctpSocketFactory()
+
+        /* Copying value set by Chrome's dcsctp_transport. */
+        const val DEFAULT_MAX_TIMER_DURATION = 3000L
+
+        val DEFAULT_SOCKET_OPTIONS = DcSctpOptions().apply {
+            maxTimerBackoffDuration = DEFAULT_MAX_TIMER_DURATION
+            // TODO add a config options for this.
+            zeroChecksumAlternateErrorDetectionMethod =
+                DcSctpOptions.ZeroChecksumAlternateErrorDetectionMethod_LowerLayerDtls
+        }
+
+        val DEFAULT_SEND_OPTIONS = SendOptions()
+
+        const val DEFAULT_SCTP_PORT: Int = 5000
+    }
+}
+
+abstract class DcSctpBaseCallbacks(
+    val transport: DcSctpTransport,
+    val clock: Clock = Clock.systemUTC()
+) : DcSctpSocketCallbacks {
+    /* Methods we can usefully implement for every JVB socket */
+    override fun createTimeout(p0: DcSctpSocketCallbacks.DelayPrecision): Timeout {
+        return ATimeout()
+    }
+
+    override fun Now(): Instant {
+        return clock.instant()
+    }
+
+    override fun getRandomInt(low: Long, high: Long): Long {
+        return ThreadLocalRandom.current().nextLong(low, high)
+    }
+
+    /* Methods we wouldn't normally expect to be called for a JVB SCTP socket. */
+    override fun OnConnectionRestarted() {
+        transport.logger.info("Surprising SCTP callback: connection restarted")
+    }
+
+    override fun OnStreamsResetFailed(outgoingStreams: ShortArray, reason: String) {
+        transport.logger.info(
+            "Surprising SCTP callback: streams ${outgoingStreams.joinToString()} reset failed: $reason"
+        )
+    }
+
+    override fun OnStreamsResetPerformed(outgoingStreams: ShortArray) {
+        transport.logger.info("Surprising SCTP callback: outgoing streams ${outgoingStreams.joinToString()} reset")
+    }
+
+    override fun OnIncomingStreamsReset(incomingStreams: ShortArray) {
+        /* Does Chrome ever reset streams? */
+        transport.logger.info("Surprising SCTP callback: incoming streams ${incomingStreams.joinToString()} reset")
+    }
+
+    private inner class ATimeout : Timeout, Runnable {
+        private var timeoutId: Long = 0
+        private var future: ScheduledFuture<*>? = null
+        override fun start(duration: Long, timeoutId: Long) {
+            this.timeoutId = timeoutId
+            future = TaskPools.SCHEDULED_POOL.schedule(this, duration, TimeUnit.MILLISECONDS)
+        }
+
+        override fun stop() {
+            future?.cancel(false)
+        }
+
+        override fun run() {
+            transport.socket.handleTimeout(timeoutId)
```

Comment:
Is this safe to run in the single-threaded SCHEDULED_POOL? Perhaps keep the timer there but move the work to an IO thread?